src/libcore/ffi.rs

```diff
@@ -0,0 +1,165 @@
+
+/// A wrapper for a `va_list`
+#[lang = "va_list"]
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct VaList<'a>(VaListTy<'a>);
+
+// The VaArgSafe trait needs to be used in public interfaces, however, the trait
+// itself must not be allowed to be used outside this module. Allowing users to
+// implement the trait for a new type (thereby allowing the va_arg intrinsic to
+// be used on a new type) is likely to cause undefined behavior.
+//
+// FIXME(dlrobertson): In order to use the VaArgSafe trait in a public interface
+// but also ensure it cannot be used elsewhere, the trait needs to be public
+// within a private module. Once RFC 2145 has been implemented look into
+// improving this.
+mod sealed_trait {
+    /// Trait which whitelists the allowed types to be used with [VaList::arg]
+    ///
+    /// [VaList::va_arg]: struct.VaList.html#method.arg
+    pub trait VaArgSafe {}
+}
+
+macro_rules! impl_va_arg_safe {
+    ($($t:ty),+) => {
+        $(
+            impl sealed_trait::VaArgSafe for $t {}
+        )+
+    }
+}
+
+impl_va_arg_safe!{i8, i16, i32, i64, usize}
+impl_va_arg_safe!{u8, u16, u32, u64, isize}
+impl_va_arg_safe!{f64}
+
+impl<T> sealed_trait::VaArgSafe for *mut T {}
+impl<T> sealed_trait::VaArgSafe for *const T {}
+
+impl<'a> VaList<'a> {
+    #[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
+                  not(target_arch = "x86_64")),
+              windows))]
+    unsafe fn raw_mut_ptr(&mut self) -> *mut i8 {
+        &mut self.0 as *mut _ as *mut i8
+    }
+
+    #[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc",
+                  target_arch = "x86_64"),
+              not(windows)))]
+    unsafe fn raw_mut_ptr(&mut self) -> *mut i8 {
+        self.0 as *mut _ as *mut i8
+    }
+
+    /// Advance to the next arg.
+    pub unsafe fn arg<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        va_arg(self.raw_mut_ptr())
+    }
+
+    /// Copy the `va_list` at the current location.
+    pub unsafe fn copy<F, R>(&mut self, f: F) -> R
+            where F: for<'copy> FnOnce(VaList<'copy>) -> R {
+        #[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
+                      not(target_arch = "x86_64")),
+                  windows))]
+        let ap_inner = ::mem::uninitialized::<VaListTy>();
+        #[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc",
+                      target_arch = "x86_64"),
+                  not(windows)))]
+        let mut ap_impl = ::mem::uninitialized::<VaListImpl>();
+        #[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc",
+                      target_arch = "x86_64"),
+                  not(windows)))]
+        let ap_inner = &mut ap_impl;
+
+        let mut ap = VaList(ap_inner);
+        let ap_ptr = ap.raw_mut_ptr();
+        va_copy(ap_ptr, self.raw_mut_ptr());
+        let ret = f(ap);
+        va_end(ap_ptr);
+        ret
+    }
+}
+
+extern "rust-intrinsic" {
```

Comment:
We loose some type safety here by using a raw pointer as the type of the input, but this makes the codegen a little less complex. We can either have a `raw_mut_ptr` function implemented for pointer variants of the `VaList` e.g. Windows, i686, etc or we can pass a `&mut VaList` to the intrinsics and place a conditional in the codegen that derefs the arg if the variant is a well defined structure e.g. AArch64, x86_64, etc.